### DISCUSSION TITLE
I've got some questions.

### DISCUSSION THREAD

**Discussion by shinkuan** [smly/mjai.app#44] [Q&A] [answered]
1. What does `Bot.tiles_seen()` and `Bot.forbidden_tiles()` do?
https://github.com/smly/mjai.app/blob/b53fa40e645313b9eba861d2fbcf38bf7f32f19e/python/mjai/bot/base.py#L347-L357
2. Is there a way that I can know which pai I can discard after I know `Bot.can_discard == True`
3. Same as 2 but for riichi.

**Comment by smly** [comment, marked as answer]
> 1. What does Bot.tiles_seen() and Bot.forbidden_tiles() do?

I updated the description for those properties.
https://github.com/smly/mjai.app/blob/1909e02fd005f780bc302458481c3a8bfab3c487/python/mjai/bot/base.py#L348-L380

> 2 & 3

Good question. For now, there are no way to know those tiles directly.

As for (2), the discardable tiles can be determined by `self.tehai_mjai` and `self.forbidden_tiles`.

```py
discardable_tiles = [tile for tile in self.tehai_mjai if not self.forbidden_tiles[tile]]
```

(3) is to find discard tiles that makes the player's hand tenpai.

```py
# NOTE: The function names are not good. These names will be changed in a few hours.
from mjai.mlibriichi.tools import calc_shanten
from mjai.bot.tools import convert_tehai_vec34_as_short, convert_mjai_tiles_to_vec34

# Just after riichi declaration
discardable_tiles = [
    tile for idx, tile in enumerate(self.tehai_mjai)
    if calc_shanten(convert_tehai_vec34_as_short(convert_mjai_tiles_to_vec34(
        self.tehai_mjai[:idx] + self.tehai_mjai[idx+1:]))) == 0
]
```

I will consider adding these properties. Thanks!

**Reply by smly** [reply]
↑ Added those properties on https://github.com/smly/mjai.app/pull/49
